### PR TITLE
WASM: Implement WASMAssembler Class

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
fixes https://github.com/lfortran/lfortran/issues/589

Also towards https://github.com/lcompilers/lpython/issues/1485

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please review and share feedback.

**Comment by Shaikh-Ubaid** [comment]
`X86Assembler` is present in `LCompilers` namespace (not in `X86` namespace). Therefore, this commit [WASM: Bring WASMAssembler outside of wasm namespace](https://github.com/lcompilers/lpython/pull/1675/commits/5fcf60fbe76c49970e1ce1a988ff6dd35328ccb5).

**Comment by Shaikh-Ubaid** [comment]
This PR might have conflicts with the sync PR. I will rebase (and/or resolve any conflicts) after completion/merging of the sync PR.

**Comment by czgdp1807** [comment]
> This PR might have conflicts with the sync PR. I will rebase (and/or resolve any conflicts) after completion/merging of the sync PR.

Marking it for draft until then.

**Comment by Shaikh-Ubaid** [comment]
>Marking it for draft until then.

Sure, Thank you!

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please review and share feedback.

File: src/libasr/codegen/wasm_assembler.h [hunk 1]
```diff
@@ -244,4 +86,302 @@ void save_bin(Vec<uint8_t> &code, std::string filename) {
...
+    Allocator &m_al;
+
+    Vec<uint8_t> m_type_section;
+    Vec<uint8_t> m_import_section;
+    Vec<uint8_t> m_func_section;
+    Vec<uint8_t> m_memory_section;
+    Vec<uint8_t> m_global_section;
+    Vec<uint8_t> m_export_section;
+    Vec<uint8_t> m_code_section;
+    Vec<uint8_t> m_data_section;
```
Comment (certik) [inline src/libasr/codegen/wasm_assembler.h:100]: I think this part should go into WASMFormat class. The WASMAssembler should just create the `m_code_section` part, the rest should be handled by WASMFormat.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_assembler.h:100]: The `WASMFormat` Class is named as `WASMAssembler` here. The `WASMAssembler` which just creates the `m_code_section` is named as `WASMInstsAssembler`. I will update the names if these are confusing.

Currently, the `WASMAssembler` class inherits from `WASMInstsAssembler` thereby  obtaining/having all the features/powers of creating the wasm format as well as the wasm instructions.

Comment (certik) [reply src/libasr/codegen/wasm_assembler.h:100]: Yes, the names are confusing to me. I would keep them completely separate no inheritance. The "assembler" knows how to convert instructions to a binary representation. The "binary format" knows how to create sections, headers, etc. and how to create the final binary. It just copies over the result from the "assembler". If you want, we can discuss this at our meeting more and see what would be a good design.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_assembler.h:100]: >If you want, we can discuss this at our meeting more and see what would be a good design.

Sure Sir, let us discuss it in the meet.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_assembler.h:100]: I updated the above comment https://github.com/lcompilers/lpython/pull/1675#discussion_r1161315091.

**Review by certik — APPROVED** [review]
I think this is fine. We have the "intruction assembler" class that is automatically generated and handles just converting instructions to binary, into the code section (only). Then we have the "full assembler" class that is built on top and handles everything else and emits the whole WASM binary format, which contains various sections and other parts that must work well together.

**Pull request merged**